### PR TITLE
[#154345339][#155241059] Feature chat mentions upgrade

### PR DESCRIPTION
**Pull request by armandgray**
Functionality: 
1. Tracks the start and end indices of any and all @ mentions in the given textInput.
2. Deletes any mention if the last char of the mention in deleted and there is a space after it or the cursor moved away from the mention.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -65,70 +63,355 @@ export class MentionsTextInput extends Component {
     }).start();
   }
 
-  isTriggerDeleted(cursor) {
-    return false;
+  handleDisplaySuggestions(position) {
+      if (!this.triggerMatrix || !this.triggerMatrix.length || !this.isTrackingStarted || this.shouldDeleteTriggerOnBackspace) {
```
Comment (MattFoley) [inline index.js]: Too many conditionals on one line.

File: index.js [hunk 2]
```diff
@@ -65,70 +63,355 @@ export class MentionsTextInput extends Component {
...
+  handleDeleteTriggerOnBackspace(position = 0, index = -2) {
+    if (!this.triggerMatrix || !this.triggerMatrix.length) {
+      return;
+    }
+
+    if (index === -2) {
+      index = this.getSubsequentTriggerIndex(position);
+    }
+
+    const isAtEnd = position === this.state.text.length - 1;
```
Comment (MattFoley) [inline index.js:99]: Do you reuse any of these logic components?

File: index.js [hunk 3]
```diff
@@ -65,70 +63,355 @@ export class MentionsTextInput extends Component {
...
-        this.closeSuggestionsPanel();
+  isTriggerSplitBySpace(position) {
+    if (!this.triggerMatrix || !this.triggerMatrix.length) {
+      return false;
+    }
+
+    const index = this.getSubsequentTriggerIndex(position);
+    return this.isPositionWithinTrigger(position, index)
+              && this.isTrackingStarted
+              && this.state.text[position] === ' ';
```
Comment (MattFoley) [inline index.js:148]: Maybe try to get in the habit of using `==` instead of `===` everywhere.

File: index.js [hunk 4]
```diff
@@ -65,70 +63,355 @@ export class MentionsTextInput extends Component {
...
+      return;
+    }
+
+    const start = this.triggerMatrix[index][0];
+    const end = this.triggerMatrix[index][1];
+    const pattern = new RegExp(`${this.props.trigger}[a-z0-9_-]*`, `gi`);
+    const triggerText = this.state.text.slice(start, end + this.getDistanceToNextSpace(end) + 1);
+    const keywordArray = triggerText.match(pattern);
+
+    return keywordArray && !!keywordArray.length ? keywordArray[0] : '';
```
Comment (MattFoley) [inline index.js]: Why `!!` vs just `keywordArray.length`?

File: index.js (hunk 2 above)
Comment (armandgray) [reply index.js:99]: there like that for naming/ readability... makes it a lot easier to follow the return value

File: index.js (hunk 1 above)
Comment (armandgray) [reply index.js]: 👍

File: index.js (hunk 3 above)
Comment (armandgray) [reply index.js:148]: wait i thought it was the opposite?

Comment (armandgray) [reply index.js:148]: 👍

File: index.js (hunk 4 above)
Comment (armandgray) [reply index.js]: its the strict true false evaluation of the length. just let it like that from the original repo. going to change it

Comment (MattFoley) [reply index.js]: I'm fine with that then.

**Review by MattFoley — APPROVED** [review]
👍

File: index.js (hunk 3 above)
Comment (gabrielfirmacion) [reply index.js:148]: I also thought it was the opposite 😕

**Review by gabrielfirmacion — APPROVED** [review]
💯

**Pull request merged**